marimo/_islands/island_generator.py

```diff
@@ -0,0 +1,189 @@
+
+        *Returns:*
+
+        - App: The built app.
+        """
+
+        if self.has_run:
+            raise ValueError("You can only call build() once")
+
+        self._session = await run_app_until_completion(
+            file_manager=AppFileManager.from_app(self._app),
+            cli_args={},
+        )
+        self.has_run = True
+        return cast(App, self._app)
+
+    def render_header(self, version_override: str = __version__) -> str:
+        """
+        Render the header for the app.
+        This should be included in the <head> tag of the page.
+        """
+
+        # This loads:
+        # - The marimo islands js
+        # - The marimo islands css
+        # - Preconnects to Google Fonts (https://stackoverflow.com/questions/73838138)
+        # - Fonts from Google Fonts
+        #   (otherwise they would get bundled in the css)
+        # - Fonts from KaTeX
+        #   (otherwise they would get bundled in the css)
+
+        base_url = f"https://cdn.jsdelivr.net/npm/@marimo-team/islands@{version_override}"
+        # This should be kept in sync fonts.css in the frontend
+        # Since this is embedded on other pages, we want display=swap
+        # for the most compatible font loading
+        font_url = "https://fonts.googleapis.com/css2?family=Fira+Mono:wght@400;500;700&amp;family=Lora&amp;family=PT+Sans:wght@400;700&amp;display=swap"
+
+        return dedent(
+            f"""
+            <script src="{base_url}/dist/main.js"></script>
+            <link
+                href="{base_url}/dist/styles.css"
+                rel="stylesheet"
+                crossorigin="anonymous"
+            />
+            <link rel="preconnect" href="https://fonts.googleapis.com" />
+            <link
+                rel="preconnect"
+                href="https://fonts.gstatic.com"
+                crossorigin
+            />
+            <link href="{font_url}" rel="stylesheet" />
+            <link
+                rel="stylesheet"
+                href="https://cdn.jsdelivr.net/npm/katex@0.16.10/dist/katex.min.css"
+                integrity="sha384-wcIxkf4k558AjM3Yz3BBFQUbk/zgIYC2R0QpeeYb+TwlBVMrlgLqwRjRtGZiK7ww"
+                crossorigin="anonymous"
+            />
+            """
+        ).strip()
+
+    def render(self, code: str) -> str:
+        """
+        Render the HTML code for the given Python code.
+        """
+
+        if not self.has_run:
+            raise ValueError("You must call build() before calling render()")
+
+        # Find the cell that has the code
+        cell_id = self._app.cell_manager.get_cell_id_by_code(code)
+        if cell_id is None:
+            LOGGER.warning(f"Could not find cell with code: {code}")
+            return ""
+
+        outputs = self._session.session_view.get_cell_outputs([cell_id])
+        output = outputs.get(cell_id, None)
+
+        # Cell may not have output
+        # (e.g. imports, but still needs to be included)
+
+        return dedent(
```

Comment:
Figures (images) and tables have to be handled specifically for exports to PDF (HTML, much less react components don't render)

Not saying this needs to be handled here, but food for thought.
I think putting render on the stub class and then exposing "output" could address those cases